aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -318,35 +277,67 @@ static void isin_sorting(
   }
 }
 
-Tensor where(const Tensor& condition, const Tensor& self, const Tensor& other) {
-  TORCH_CHECK(condition.device() == self.device() && self.device() == other.device(),
-              "Expected condition, x and y to be on the same device, but condition is on ",
-              condition.device(), " and x and y are on ", self.device(), " and ", other.device(),
-              " respectively");
+namespace {
+Tensor _s_where_dispatch(
+    const Tensor& condition,
+    const Tensor& self,
+    const Tensor& other) {
+  TORCH_CHECK(
+      condition.device() == self.device() && self.device() == other.device(),
+      "Expected condition, x and y to be on the same device, but condition is on ",
+      condition.device(),
+      " and x and y are on ",
+      self.device(),
+      " and ",
+      other.device(),
+      " respectively");
 
   if (condition.scalar_type() == ScalarType::Byte) {
-  TORCH_WARN_ONCE("where received a uint8 condition tensor. This behavior is deprecated and will be removed in a future version of PyTorch. Use a boolean condition instead.");
-} else {
-  TORCH_CHECK(condition.scalar_type() == ScalarType::Bool, "where expected condition to be a boolean tensor, but got a tensor with dtype ", condition.scalar_type());
-}
+    TORCH_WARN_ONCE(
+        "where received a uint8 condition tensor. This behavior is deprecated and will be removed in a future version of PyTorch. Use a boolean condition instead.");
+  } else {
+    TORCH_CHECK(
+        condition.scalar_type() == ScalarType::Bool,
+        "where expected condition to be a boolean tensor, but got a tensor with dtype ",
+        condition.scalar_type());
+  }
 
   c10::MaybeOwned<Tensor> b_condition, b_self, b_other;
-  std::tie(b_condition, b_self, b_other) = expand_outplace(condition, self, other, "where");
+  std::tie(b_condition, b_self, b_other) =
+      expand_outplace(condition, self, other, "where");
   return at::_s_where(*b_condition, *b_self, *b_other);
 }
 
+Tensor get_promoted_tensor(const Tensor& t, ScalarType common_dtype) {
+  return (t.device().is_cpu()) && !(t.scalar_type() == common_dtype)
+      ? t.to(common_dtype)
+      : t;
+}
+
+} // namespace
+
+Tensor where(const Tensor& condition, const Tensor& self, const Tensor& other) {
+  TORCH_CHECK(self.dtype() == other.dtype(), "expected scalar type ", self.dtype(), " but found ", other.dtype());
+  return _s_where_dispatch(condition, self, other);
+}
+
 Tensor where(const Tensor& condition, const Scalar& self, const Tensor& other) {
-  return at::where(condition, wrapped_scalar_tensor(self, other.device()), other);
+  auto common_dtype = at::result_type(self, other);
```

Comment:
Let's add a where type promotion note and reference it in places like this where the type promotion happens so readers can more easily understand what's going on